lib/zip_source_zip_new.c

```diff
@@ -100,29 +132,50 @@ zip_source_t *_zip_source_zip_new(zip_t *srcza, zip_uint64_t srcidx, zip_flags_t
     }
     _zip_file_attributes_from_dirent(&attributes, de);
 
-    if (st.comp_size == 0) {
+    if (st.size == 0) {
```

Comment:
This is wrong. You have to check if the data we want to read is 0 bytes long. If we're reading compressed data, that's st.comp_size, otherwise is st.size. If that isn't known, assume it's not 0.